### ISSUE TITLE
AsSuperVisitor: type is not an erased subtype of supertype 14

### ISSUE THREAD

**Issue by cushon** [typetools/checker-framework#1587]
```java
class One implements Two<One, Three, Four, Four> {}

interface Two<E extends Two<E, K, C, I>, K extends Five<K>, C extends Enum<C>, I extends Enum<I>> {}

abstract class Three implements Five<Three> {}

enum Four {}

interface Five<K extends Five<K>> extends Comparable<K> {}

interface Six {
  <E extends Two<E, ?, ?, I>, I extends Enum<I>> Seven<?, E> e(Class<E> entity);
}

interface Seven<T extends Seven<T, E>, E extends Two<E, ?, ?, ?>> extends Eight<T, E, E> {}

interface Eight<T extends Eight<T, R, E>, R, E extends Two<E, ?, ?, ?>> extends Nine<T, R> {}

interface Nine<T extends Nine<T, R>, R> {
  T d();

  Iterable<R> q();
}

abstract class T {
  public Iterable<One> f(Six e) {
    return g(e.e(One.class).d().q());
  }

  abstract Iterable<One> g(Iterable<One> r);
}
```

```
$ ~/jsr308/checker-framework-2.2.1/checker/bin/javac -processor Nullness T.java -AprintErrorStack
...
error: AsSuperVisitor: type is not an erased subtype of supertype.
  type: Two<E extends Two<E, ? extends Five<K extends Five<K extends Five<K>>>, ? extends Enum<C extends Enum<C extends Enum<C>>>, ? extends Enum<I extends Enum<I extends Enum<I>>>>, ? extends Five<K extends Five<K extends Five<K>>>, ? extends Enum<C extends Enum<C extends Enum<C>>>, ? extends Enum<I extends Enum<I extends Enum<I>>>>
  superType: One
  Compilation unit: /Users/cushon/src/error-prone/core/src/main/java/com/google/errorprone/bugpatterns/X.java
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:727)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:41)
  org.checkerframework.framework.type.AsSuperVisitor.errorTypeNotErasedSubtypeOfSuperType(AsSuperVisitor.java:155)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:340)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:29)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:308)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:59)
...
error: AsSuperVisitor: type is not an erased subtype of supertype.
  type: @Initialized @NonNull Two<E extends @Initialized @NonNull Two<E, ? extends @Initialized @NonNull Five<K extends @Initialized @NonNull Five<K extends @Initialized @NonNull Five<K>>>, ? extends @Initialized @NonNull Enum<C extends @Initialized @NonNull Enum<C extends @Initialized @NonNull Enum<C>>>, ? extends @Initialized @NonNull Enum<I extends @Initialized @NonNull Enum<I extends @Initialized @NonNull Enum<I>>>>, ? extends @Initialized @NonNull Five<K extends @Initialized @NonNull Five<K extends @Initialized @NonNull Five<K>>>, ? extends @Initialized @NonNull Enum<C extends @Initialized @NonNull Enum<C extends @Initialized @NonNull Enum<C>>>, ? extends @Initialized @NonNull Enum<I extends @Initialized @NonNull Enum<I extends @Initialized @NonNull Enum<I>>>>
  superType: @Initialized @NonNull One
  Compilation unit: /Users/cushon/src/error-prone/core/src/main/java/com/google/errorprone/bugpatterns/X.java
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:727)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:41)
  org.checkerframework.framework.type.AsSuperVisitor.errorTypeNotErasedSubtypeOfSuperType(AsSuperVisitor.java:155)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:340)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:29)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:308)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:59)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:86)
  org.checkerframework.framework.type.AsSuperVisitor.visitTypevar_NotTypevarNorWildcard(AsSuperVisitor.java:586)
```

**Comment by smillst** [comment]
Thanks for reporting this!  I was able to further reduce the test case to:
```java
abstract class Issue1587 {
    static class MyObject {}
    interface Six<T extends Six<T, R>, R> {
        T d();
        Iterable<R> q();
    }
    abstract Six<?, MyObject> e(Object entity);
    public void method(MyObject one) {
        g(e(one).d().q());
    }
    abstract Iterable<MyObject> g(Iterable<MyObject> r);
}
```
(But we should add the original as a test case, too.)

**Comment by smillst** [comment]
Fixed via feedb6119b9c76098c7f
